### PR TITLE
ASIM DNS schema parser with its sample and test data for SentinelOne

### PR DESCRIPTION
**Pull request by jayeshprajapaticrest**
Required items, please complete
   
   Change(s):
   - Added ASIM DNS schema parser for SentinelOne based on API Data

   Reason for Change(s):
   - Initial version ASIM DNS Parser

   Version Updated:
   - Initial version

   Testing Completed:
   - Yes

   Checked that the validations are passing and have addressed any issues that are present:
   - Yes

### CODE REVIEW

**Comment by v-rusraut** [comment]
Hi @jayeshprajapaticrest,
Thank you for raising Pull Request with us! We will review it internally and get back to you by shortly.

**Review by vakohl — CHANGES_REQUESTED** [review]

**Comment by v-sudkharat** [comment]
Hi @jayeshprajapaticrest, hope you are doing well. Just wanted to check if you got a chance to look at the suggestions shared. Please feel free to reach out to us for any queries or support. Thanks!

**Comment by jayeshprajapaticrest** [comment]
> Hi @jayeshprajapaticrest, hope you are doing well. Just wanted to check if you got a chance to look at the suggestions shared. Please feel free to reach out to us for any queries or support. Thanks!

@v-sudkharat We will look into above suggested changes and revert here as soon as possible.

File: Parsers/ASimDns/Parsers/ASimDnsSentinelOne.yaml [hunk 1]
```diff
@@ -0,0 +1,148 @@
+Parser:
+  Title: DNS activity ASIM parser for SentinelOne
+  Version: '0.1.1'
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimDns/Parsers/ASimDnsSentinelOne.yaml]: Done

File: Parsers/ASimDns/Parsers/ASimDnsSentinelOne.yaml [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+- Title: ASIM DNS Schema
+  Link: https://aka.ms/ASimDnsDoc
+- Title: ASIM
+  Link: https://aka.ms/AboutASIM
+- Title: SentinelOne Documentation
+  Link: https://<SOneInstanceDomain>.sentinelone.net/api-doc/overview
+Description: |
+  This ASIM parser supports normalizing SentinelOne logs to the ASIM DNS normalized schema. SentinelOne events are captured through SentinelOne data connector which ingests SentinelOne server objects such as Threats, Agents, Applications, Activities, Policies, Groups, and more events into Microsoft Sentinel through the REST API.
+ParserName: ASimDnsSentinelOne
+EquivalentBuiltInParser: _Im_DNS_SentinelOne
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimDns/Parsers/ASimDnsSentinelOne.yaml]: Done

File: Parsers/ASimDns/Parsers/ASimDnsSentinelOne.yaml [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+ParserParams:
+  - Name: disabled
+    Type: bool
+    Default: false
+
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimDns/Parsers/ASimDnsSentinelOne.yaml]: Done

File: Parsers/ASimDns/Parsers/ASimDnsSentinelOne.yaml [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+Normalization:
+  Schema: Dns
+  Version: '0.1.7'
+References:
+- Title: ASIM DNS Schema
+  Link: https://aka.ms/ASimDnsDoc
+- Title: ASIM
+  Link: https://aka.ms/AboutASIM
+- Title: SentinelOne Documentation
+  Link: https://<SOneInstanceDomain>.sentinelone.net/api-doc/overview
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimDns/Parsers/ASimDnsSentinelOne.yaml:16]: As discussed the SentinelOne doesn't have the public documentation available for API.
User can refer from their instance URL.

File: Parsers/ASimDns/Parsers/ASimDnsSentinelOne.yaml [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+      | extend
+          Dvc = DvcId,
+          EventEndTime = EventStartTime,
+          EventResult = iff(DnsResponseCode == 0, "Success", "Failure"),
+          EventResultDetails = _ASIM_LookupDnsResponseCode(DnsResponseCode),
+          DnsQueryTypeName = _ASIM_LookupDnsQueryType(DnsQueryType),
+          Rule = RuleName,
+          SrcDvcId = DvcId,
+          SrcHostname = DvcHostname,
+          EventSeverity = iff(EventOriginalSeverity == "Critical", "High", EventOriginalSeverity),
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimDns/Parsers/ASimDnsSentinelOne.yaml:117]: Yes

File: Parsers/ASimDns/Parsers/vimDnsSentinelOne.yaml [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+                        agentDetectionInfo_osRevision_s
+                    )
+      | extend EventResultDetails = _ASIM_LookupDnsResponseCode(DnsResponseCode)
+      | where (responsecodename == '*' or EventResultDetails =~ responsecodename)
+      | extend 
+          DnsQueryType = iff(isempty(DnsQueryType) and DnsResponseCode == 0, int(1), DnsQueryType)
+      | project-rename
+          EventStartTime = sourceProcessInfo_pidStarttime_t,
+          DnsQuery = alertInfo_dnsRequest_s,
+          EventUid = _ResourceId,
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimDns/Parsers/vimDnsSentinelOne.yaml]: Done

File: Parsers/ASimDns/Parsers/vimDnsSentinelOne.yaml [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+      eventtype: string='Query',
+      disabled: bool=false
+      ) {
+      SentinelOne_CL
+      | where not(disabled)
+          and (eventtype == '*' or eventtype == "Query")
+          and (isnull(starttime) or TimeGenerated >= starttime)
+          and (isnull(endtime) or TimeGenerated <= endtime)
+          and event_name_s == "Alerts." 
+          and alertInfo_eventType_s == "DNS"
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimDns/Parsers/vimDnsSentinelOne.yaml:67]: Added

File: Parsers/ASimDns/test/SentinelOne_ASimDns_SchemaTest.csv [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+"(2) Info: Missing optional field [SrcRiskLevel]"
+"(2) Info: Missing optional field [SrcUserAWSId]"
+"(2) Info: Missing optional field [SrcUserAadId]"
+"(2) Info: Missing optional field [SrcUserId]"
+"(2) Info: Missing optional field [SrcUserOktaId]"
+"(2) Info: Missing optional field [SrcUserScopeId]"
+"(2) Info: Missing optional field [SrcUserScope]"
+"(2) Info: Missing optional field [SrcUserSessionId]"
+"(2) Info: Missing optional field [SrcUserSid]"
+"(2) Info: Missing optional field [SrcUserType]"
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimDns/test/SentinelOne_ASimDns_SchemaTest.csv]: As discussed we have added the SrcUserType field by providing an empty string for userid parameter in _ASIM_GetUserType().

File: Parsers/ASimDns/Parsers/vimDnsSentinelOne.yaml [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+      | extend EventResultDetails = _ASIM_LookupDnsResponseCode(DnsResponseCode)
+      | where (responsecodename == '*' or EventResultDetails =~ responsecodename)
+      | extend 
+          DnsQueryType = iff(isempty(DnsQueryType) and DnsResponseCode == 0, int(1), DnsQueryType)
+      | project-rename
+          EventStartTime = sourceProcessInfo_pidStarttime_t,
+          DnsQuery = alertInfo_dnsRequest_s,
+          EventUid = _ResourceId,
+          DvcHostname = agentDetectionInfo_name_s,
+          DnsResponseName = alertInfo_dnsResponse_s,
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimDns/Parsers/vimDnsSentinelOne.yaml:137]: Done.

File: Parsers/ASimDns/test/SentinelOne_ASimDns_SchemaTest.csv [hunk 10]
```diff
@@ -0,0 +1,101 @@
+﻿Result
+"(1) Warning: Missing recommended field [Dst]"
+"(1) Warning: Missing recommended field [DvcDomain]"
+"(1) Warning: Missing recommended field [DvcIpAddr]"
+"(1) Warning: Missing recommended field [SrcDomain]"
+"(1) Warning: Missing recommended field [SrcIpAddr]"
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimDns/test/SentinelOne_ASimDns_SchemaTest.csv:5]: As discussed we don't have a field related to SrcIpAddr in SentinelOne.

File: Parsers/ASimDns/Parsers/ASimDnsSentinelOne.yaml [hunk 11]
```diff
@@ -0,0 +1,148 @@
...
+          EventOriginalUid = alertInfo_dvEventId_s,
+          RuleName = ruleInfo_name_s,
+          SrcProcessId = sourceProcessInfo_pid_s,
+          SrcProcessName = sourceProcessInfo_name_s,
+          SrcUsername = sourceProcessInfo_user_s
+      | extend
+          Dvc = DvcId,
+          EventEndTime = EventStartTime,
+          EventResult = iff(DnsResponseCode == 0, "Success", "Failure"),
+          EventResultDetails = _ASIM_LookupDnsResponseCode(DnsResponseCode),
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimDns/Parsers/ASimDnsSentinelOne.yaml:110]: Done

File: Parsers/ASimDns/test/SentinelOne_vimDns_SchemaTest.csv [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+"(2) Info: Missing optional field [DvcScopeId]"
+"(2) Info: Missing optional field [DvcScope]"
+"(2) Info: Missing optional field [DvcZone]"
+"(2) Info: Missing optional field [EventMessage]"
+"(2) Info: Missing optional field [EventOriginalSubType]"
+"(2) Info: Missing optional field [EventOriginalType]"
+"(2) Info: Missing optional field [EventOwner]"
+"(2) Info: Missing optional field [EventProductVersion]"
+"(2) Info: Missing optional field [EventReportUrl]"
+"(2) Info: Missing optional field [EventSubType]"
```
[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimDns/test/SentinelOne_vimDns_SchemaTest.csv]: As mentioned [here](https://github.com/Azure/Azure-Sentinel/pull/8666#discussion_r1290343784), we have set EvenSubType to Response when value is available in the field alertInfo_dnsResponse_s else set to Request.

File: Parsers/ASimDns/Parsers/ASimDnsSentinelOne.yaml (hunk 4 above)
[the remark this answers is not in the document]
Comment (vakohl) [reply Parsers/ASimDns/Parsers/ASimDnsSentinelOne.yaml:16]: @jayeshprajapaticrest 
is there a specific page for DNS on this API page that we can give? the current one seems genetic page

File: Parsers/ASimDns/test/SentinelOne_ASimDns_SchemaTest.csv (hunk 10 above)
[the remark this answers is not in the document]
Comment (vakohl) [reply Parsers/ASimDns/test/SentinelOne_ASimDns_SchemaTest.csv:5]: @jayeshprajapaticrest 
this is an important field in DNS, can you re-check and confirm?

[the remark this answers is not in the document]
Comment (jayeshprajapaticrest) [reply Parsers/ASimDns/test/SentinelOne_ASimDns_SchemaTest.csv:5]: We double checked but there were no field which we can map with SrcIpAddr

**Comment by niralishah-crest** [comment]
@vakohl we have mapped the inspection fields below in the DNS parser, which we found from SentinelOne Alerts logs.
RuleName
Rule
ThreatConfidence
ThreatOriginalConfidence

File: Parsers/ASimDns/Parsers/vimDnsSentinelOne.yaml [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+          DvcOs = agentDetectionInfo_osName_s,
+          DvcOsVersion = agentDetectionInfo_osRevision_s,
+          EventOriginalType = alertInfo_eventType_s,
+          EventOriginalSeverity = ruleInfo_severity_s,
+          EventOriginalUid = alertInfo_dvEventId_s,
+          RuleName = ruleInfo_name_s,
+          SrcProcessId = sourceProcessInfo_pid_s,
+          SrcProcessName = sourceProcessInfo_name_s,
+          SrcUsername = sourceProcessInfo_user_s,
+          ThreatOriginalConfidence = ruleInfo_treatAsThreat_s
```
Comment (vakohl) [inline Parsers/ASimDns/Parsers/vimDnsSentinelOne.yaml:189]: @jayeshprajapaticrest @niralishah-crest 
Hope you have validated with all available ruleInfo_treatAsThreat_s values. In the sample logs, I only had 'UNDEFINED' value

File: Parsers/ASimDns/Parsers/ASimDnsSentinelOne.yaml [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+  ];
+  let ThreatConfidenceLookup_suspicious = datatable(
+      alertInfo_analystVerdict_s: string,
+      ThreatConfidence_suspicious: int
+  )
+  [
+      "FALSE_POSITIVE", 40,
+      "Undefined", 50,
+      "SUSPICIOUS", 60,
+      "TRUE_POSITIVE", 67
```
Comment (vakohl) [inline Parsers/ASimDns/Parsers/ASimDnsSentinelOne.yaml:44]: any specific reason you kept this as 67 and not higher value incase of Suspicious TruePositive? e.g. 75

Comment (jayeshprajapaticrest) [reply Parsers/ASimDns/Parsers/ASimDnsSentinelOne.yaml:44]: As per our past experience we have defined the main category in three range as below:
- Undefined : 0~33
- Suspicious: 34~67
- Malicious: 68~100

And then assigned final confidence score based on the combination of "ruleInfo_treatAsThreat_s" and "alertInfo_analystVerdict_s" value as below:
<html>
<body>
<google-sheets-html-origin>

ruleInfo_treatAsThreat_s | alertInfo_analystVerdict_s | Confidence
-- | -- | --
UNDEFINED | FALSE_POSITIVE | 5
||Undefined | 15
||SUSPICIOUS | 25
||TRUE_POSITIVE | 33
Suspicious | FALSE_POSITIVE | 40
||Undefined | 50
||SUSPICIOUS | 60
||TRUE_POSITIVE | 67
Malicious | FALSE_POSITIVE | 75
||Undefined | 80
||SUSPICIOUS | 90
||TRUE_POSITIVE | 100


</body>
</html>

File: Parsers/ASimDns/Parsers/vimDnsSentinelOne.yaml (hunk 13 above)
Comment (jayeshprajapaticrest) [reply Parsers/ASimDns/Parsers/vimDnsSentinelOne.yaml:189]: @vakohl 
Actually we don not have the data of other type. But yes we have tested by ingesting dummy data with this values and its working as expected. Thanks

**Review by vakohl — DISMISSED** [review]

**Review by v-atulyadav — DISMISSED** [review]

**Review by v-atulyadav — APPROVED** [review]

**Pull request merged**